fem/bilininteg.hpp

```diff
@@ -3035,6 +3089,10 @@ class ElasticityIntegrator : public BilinearFormIntegrator
    virtual double ComputeFluxEnergy(const FiniteElement &fluxelem,
                                     ElementTransformation &Trans,
                                     Vector &flux, Vector *d_energy = NULL);
+
+   //This would be generally useful for these "component integrators". Starting
+   //to look like this should be a child class.
+   const FiniteElementSpace* GetFESpace() const;
```

Comment:
This interface is a little weird. I understand it is for the "component integrators", which themselves are a bit strange. I think this part of the code might be improved by some design changes.

Reply:
My thinking with the component integrators was that a monolithic operator, such as elasticity, should be able to be decomposable into it's exact components for preconditioning/solver purposes. I wanted to ensure consistency so that the monolithic (what I called `parent`, but I should have called `monolithic`) operator is exactly reproducible by a block operator of its components. Part of ensuring consistency is making sure the `FiniteElementSpace` returned by component integrators is the scalar version of the vector `FiniteElementSpace` and that the quadrature spaces, `QuadratureFunctions` or `CoefficientVector`s used in partial assembly were the same. So rather than allowing the user to accidentally use an incompatible `FiniteElementSpace` for the components, I thought it made sense to create an interface that ensured they were consistent.

Would it make sense to move some of this functionality that ensures consistency into `BilinearForm` and/or `BilinearFormExtension` somehow? In the miniapp I create a `BilinearForm` for the component integrator via

```
         auto *block = static_cast<decltype(integrator)*>
                       (lor_integrator.ComponentIntegrator(j,j));
         auto *fes_block = dynamic_cast<ParFiniteElementSpace*>
                           (const_cast<FiniteElementSpace*>
                            (block->GetFESpace()));
         bilinear_forms.emplace_back(new ParBilinearForm(fes_block));
         bilinear_forms[j]->AddDomainIntegrator(block);
```

I think it would make sense to create a simple function in `BilinearForm` that returns a component version of an existing  `BilinearForm`. I could then move `GetFESpace` somewhere else and turn `GetComponentIntegrator` into a private or protected function. I think we'd still want to share resources between the monolithic and component integrators somehow like I'm currently doing with `shared_ptr`, but that's hidden from the user and this removes a lot of manual bookkeeping.

The interface could look like
```
BilinearForm BilinearForm::GetComponent(const int I, const int J);
```
and the result is a BilinearForm on the I,J components. This would assume that all the integrators have component versions, and maybe needs to assume that they are using `byNODES` ordering. And these assumptions could be verified. Does this approach seem reasonable?

Reply:
Is there any conclusion on this? ping @pazner